common/quotas/global/collection/collection.go

```diff
@@ -35,66 +35,191 @@ package collection
 
 import (
 	"context"
+	"errors"
 	"fmt"
+	"strings"
 	"testing"
 	"time"
 
+	"golang.org/x/time/rate"
+
 	"github.com/uber/cadence/common/clock"
 	"github.com/uber/cadence/common/dynamicconfig"
 	"github.com/uber/cadence/common/log"
 	"github.com/uber/cadence/common/log/tag"
 	"github.com/uber/cadence/common/metrics"
 	"github.com/uber/cadence/common/quotas"
 	"github.com/uber/cadence/common/quotas/global/collection/internal"
+	"github.com/uber/cadence/common/quotas/global/rpc"
 )
 
 type (
+	// Collection wraps three kinds of ratelimiter collections:
+	//   1. a "global" collection, which tracks usage, and is weighted to match request patterns.
+	//   2. a "local" collection, which tracks usage, and can be optionally used (and/or shadowed) instead of "global"
+	//   3. a "disabled" collection, which does NOT track usage, and is used to bypass as much of this Collection as possible
+	//
+	// 1 is the reason this Collection exists - limiter-usage has to be tracked and submitted to aggregating hosts to
+	// drive the whole "global load-balanced ratelimiter" system.  internally, this will fall back to a local collection
+	// if there is insufficient data or too many errors.
+	//
+	// 2 is the previous non-global behavior, where ratelimits are determined locally, more simply, and load information is not shared.
+	// this is a lower-cost and MUCH less complex system, and it SHOULD be used if your Cadence cluster receives requests
+	// in a roughly random way (e.g. any client-side request goes to a roughly-fair roughly-random Frontend host).
+	//
+	// 3 disables as much of 1 and 2 as possible, and is intended to be temporary.
```

Comment:
What does 3 mean? 
- New, global rat-limiter is off
- Host-level (old behaviour) rate-limiter is off
.. ??

Does this imply that during 3 rollout we should expect there to be not rate-limiter events, and then we go from 3 to 2, then to 1?

Reply:
3 is "almost exactly current behavior".  The original `quotas.Collection` is used verbatim and the limiter it returns is untouched, so it's as close to feature-flag-disabling this whole pile of code as is reasonably possible.

2 is 3 but with counting and metrics emission (but no load-reporting requests).

so 3 basically just exists to be the panic button on the initial rollout.  it'll be replaced by 2 soonish, since the metrics seem worth gathering, unless they're cataclysmically bad somehow we didn't forsee.

Reply:
added!